gms/gossiper.cc

```diff
@@ -2087,20 +2087,20 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
         logger.debug("not replacing a previous ep_state for {}, but reusing it: {}", ep, ep_state);
         ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
     }
-    const auto tmptr = get_token_metadata_ptr();
-    auto tokens = tmptr->get_tokens(ep);
-    if (!tokens.empty()) {
-        std::unordered_set<dht::token> tokens_set(tokens.begin(), tokens.end());
-        ep_state.add_application_state(gms::application_state::TOKENS, versioned_value::tokens(tokens_set));
+    if (!loaded_state.tokens.empty()) {
+        ep_state.add_application_state(gms::application_state::TOKENS, versioned_value::tokens(loaded_state.tokens));
+    }
+    if (loaded_state.host_id) {
+        ep_state.add_application_state(gms::application_state::HOST_ID, versioned_value::host_id(loaded_state.host_id));
     }
-    auto host_id = tmptr->get_host_id_if_known(ep);
-    if (host_id) {
-        ep_state.add_application_state(gms::application_state::HOST_ID, versioned_value::host_id(host_id.value()));
+    if (loaded_state.opt_dc_rack) {
+        ep_state.add_application_state(gms::application_state::DC, versioned_value::datacenter(loaded_state.opt_dc_rack->dc));
+        ep_state.add_application_state(gms::application_state::RACK, versioned_value::datacenter(loaded_state.opt_dc_rack->rack));
     }
     auto generation = ep_state.get_heart_beat_state().get_generation();
+    logger.debug("Adding saved endpoint {} generation={} version={}: {}", ep, generation, ep_state.get_heart_beat_state().get_heart_beat_version(), ep_state);
```

Comment:
Generation will be always 0 here, right?

Reply:
If the endpoint state already exists we reuse it so generation might not be zero.
That was the existing logic.  I'm not sure if it's really required.

Reply:
If the endpoint exists, we call:
```
ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
```

It would be wrong to set a state in the current generation on behalf of another node using our local version generator, so it must be set with generation 0 for correctness.

Reply:
You're right. It should reuse `ep_state.get_heart_beat_state()` instead of the local `heart_beat_state`, as done in `apply_new_states`, so it just updates the timestamp and keeps the heartbeat_state.

I opened https://github.com/scylladb/scylladb/issues/16429 and will include the fix in the patchset so it can be backported independently if we want to.

Reply:
https://github.com/scylladb/scylladb/pull/16432